distributed/scheduler.py

```diff
@@ -3062,25 +3070,22 @@ def check_idle_saturated(self, ws: WorkerState, occ: float = -1.0):
 
         idle = self.idle
         saturated = self.saturated
-        if (
-            self.is_unoccupied(ws, occ, p)
-            if math.isinf(self.WORKER_SATURATION)
-            else not _worker_full(ws, self.WORKER_SATURATION)
-        ):
+        saturated.discard(ws)
+        if self.is_unoccupied(ws, occ, p):
             if ws.status == Status.running:
                 idle[ws.address] = ws
-            saturated.discard(ws)
         else:
             idle.pop(ws.address, None)
-
             nc = ws.nthreads
             if p > nc:
                 pending = occ * (p - nc) / (p * nc)
                 if 0.4 < pending > 1.9 * (self.total_occupancy / self.total_nthreads):
                     saturated.add(ws)
-                    return
 
-            saturated.discard(ws)
+        self.idle_task_count.pop(ws.address, None)
+        if not _worker_full(ws, self.WORKER_SATURATION):
+            if ws.status == Status.running:
+                self.idle_task_count[ws.address] = ws
```

Comment:
```suggestion
        if not _worker_full(ws, self.WORKER_SATURATION):
            if ws.status == Status.running:
                self.idle_task_count[ws.address] = ws
        else:
            self.idle_task_count.pop(ws.address, None)
```
This is equivalent to the previous. If the status is not running, it's already been removed by `handle_worker_status_change`, so we don't need to worry about that case.